src/combinatorics/hindman.lean

```diff
@@ -0,0 +1,143 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import topology.stone_cech
+import data.stream
+
+/-!
+# Hindman's theorem on finite sums
+
+We prove Hindman's theorem on finite sums, using idempotent ultrafilters.
+
+Given an infinite sequence `a₀, a₁, a₂, …` of positive integers, the set `FS(a₀, …)` is the set
+of positive integers that can be expressed as a finite sum of `aᵢ`'s, without repetition. Hindman's
+theorem asserts that whenever the positive integers are finitely colored, there exists a sequence
+`a₀, a₁, a₂, …` such that `FS(a₀, …)` is monochromatic. We prove the theorem for a general
+semigroup `M` instead of `ℕ+` since it is no harder, although this special case implies the general
+case.
+
+The idea of the proof is to extend the addition `(+) : M → M → M` to addition `(+) : βM → βM → βM`
+on the space `βM` of ultrafilters on `M`. One can prove that if `U` is an _idempotent_ ultrafilter,
+i.e. `U + U = U`, then any `U`-large subset of `M` contains some set `FS(a₀, …)` (see
+`exists_FS_of_large`). And a general topological argument shows the existence of an idempotent
+ultrafilter (see `exists_idempotent`).
+
+## Main results
+
+- `exists_FS_of_finite_cover`: Hindman's theorem on finite sums.
+
+-/
+
+open filter
+
+instance {M} [has_add M] : has_add (ultrafilter M) := { add := λ U V, (+) <$> U <*> V }
```

Comment:
Should this move to another file? Probably close to the defn of `ultrafilter`?
Also, should this be the `to_additive` of a multiplicative defn?

Reply:
I'm not sure, addition on ultrafilters is not a super-wellbehaved operation, e.g. it's non-commutative, and for filters I think there's another notion of addition which may be more useful.

Using `to_additive` is probably a good idea.